[숫자 야구 게임] 김승현 미션 제출합니다.

seunghye218 opened the pull request.

seunghye218 committed:
docs: add list of game features

seunghye218 committed:
docs: change list of game features

seunghye218 committed:
feat: init game

seunghye218 committed:
feat: handling user input during games

seunghye218 committed:
feat: game logic

seunghye218 committed:
feat: print game result

seunghye218 committed:
docs: add post game output

seunghye218 committed:
feat: add post game logic
- pass ApplicationTest

seunghye218 committed:
refactor: Separate computer logic
- add Computer class

seunghye218 committed:
refactor: Separate user input logic
- chore: move file Computer class
- add UserInput class

seunghye218 committed:
chore: add public UserInput class

seunghye218 committed:
refactor: separate game logic
- add NumberBaseballGame class

seunghye218 committed:
refactor: google style compliance
- use this keyword
- change duplicate character logic
- change Computer.getRandomNumber

seunghye218 committed:
refactor: improve scalability
- add constants

seunghye218 committed:
chore: add constants

seunghye218 committed:
refactor: improve readability

yeseong31 commented:
MVC 패턴 없이도 프로그램을 기능대로 깔끔히 구현하신 것 같습니다! 기능 단위로 메서드 이름도 잘 지으셔서 코드 이해도 쉽다고 느껴집니다.

한 가지 의견이 있다면...

```java
public void resetNumber() {
    this.numbers = getRandomNumber();
}

public int[] getResult(String userInput) {
    int[] score = new int[2];

    for (int i = 0; i < Constants.NUMBER_SIZE; ++i) {
        char n = userInput.charAt(i);
        if (this.numbers.charAt(i) == n) {
            ++score[0];
        } else if (this.numbers.contains(String.valueOf(n))) {
            ++score[1];
        }
    }
    return score;
}
```

클래스 내부에서 필드와 메서드를 `this`를 생략하더라도 컴파일러가 자동으로 `this.`를 추가하여 컴파일을 해준다고 알고 있어서, 생성자를 제외한 부분에서는 `this` 키워드를 생략해도 괜찮지 않을까 하는 생각이 있습니다!

저는 이번 1주차 미션에 MVC 패턴으로 개발을 시도했다가 프로그램의 복잡도만 증가시켰다는 생각이 많이 들었는데, 이렇게 보니까 프로그램의 목적에 따라 MVC가 강제되지 않아도 된다는 것을 깨닫게 되는 것 같습니다. 오히려 이렇게 코드를 작성하는 것이 이해하는 데에는 더 쉬운 것 같아요. 많은 것을 느끼고 갑니다ㅎㅎ

seunghye218 commented:
지역 번수와 멤버 변수의 의미를 명확히 구분하기 위해 this 키워드를 사용했는데요!
간단한 프로그램이라 변수 개수가 적어 없는게 간결할 수 있다고 생각이 드네요. 좋은 피드백 감사드립니다~!

mijien0179 commented on src/main/java/baseball/NumberBaseballGame.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public void gameLoop() {
+        while (true) {
+            System.out.print("숫자를 입력해주세요 : ");
+            int[] score = this.computer.getResult(UserInput.getUserNumber());
+            printGameResult(score);
+            if (gameOver(score)) {
+                break;
+            }
+        }
```
`break`구문도 좋지만 다음엔 `do-while`을 이용해보는 건 어떨까요?

seunghye218 replied:
do-while 문을 거의 사용하지 않다보니 잊고있었네요!

mijien0179 commented on src/main/java/baseball/NumberBaseballGame.java:
```diff
@@ -0,0 +1,56 @@
+            result.append('볼');
+        }
+        if (score[0] != 0) {
+            if (!result.isEmpty()) {
+                result.append(' ');
+            }
```
해당 부분은 `볼`이 존재할 때, 띄어쓰기를 추가하기 위함으로 생각됩니다. 개인적인 의견이지만 볼을 출력할 때 문자열로 `"볼 "`을 출력했다면 출력상에는 큰 이상이 없으면서도 해당 구문은 삭제할 수 있었을거란 생각이 들어요.

mijien0179 commented on src/main/java/baseball/NumberBaseballGame.java:
```diff
@@ -0,0 +1,56 @@
+            System.out.println(Constants.NUMBER_SIZE + "개의 숫자를 모두 맞히셨습니다! 게임 종료");
+            System.out.println("게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.");
+            if (UserInput.getUserRestart()) {
+                this.computer.resetNumber();
```
리셋의 위치를 게임이 종료될 때가 아니라 게임의 시작 위치에서 했다면 좀 더 좋았을거란 생각이 들어요. 그렇게 되면 현재 코드상으론 첫 실행에서 숫자가 한번 변경되는 방식이지만, 실행상의 문제 없이 게임의 종료와 초기화를 분리할 수 있으니까요.

mijien0179 commented on src/main/java/baseball/UserInput.java:
```diff
@@ -0,0 +1,40 @@
+
+    static boolean getUserRestart() {
+        String restart = readLine().trim();
+        if (!(restart.equals("1") || restart.equals("2"))) {
```
리스트를 생성해서 허용 문자열을 관리한다면 더 좋을것 같다는 생각이 들어요.

```java
List<String> allowedList = Arrays.asList("1", "2");

if(!allowedList.contains(restart)){
    throw IllegalArgumentException();
}
```

seunghye218 replied:
하드코딩한 부분을 다 해결하지 못 했었네요. 많이 배워갑니다!

seunghye218 committed:
refactor: use do while statements

seunghye218 committed:
docs: add game description

seunghye218 committed:
refactor: improve code readability and maintainability

- NumberBaseballGame -> Application
- separate UserInput -> InputView, OutputView
- int[] score -> Score class
- add random number generator
- add random number validation

seunghye218 committed:
test(Computer): add test

- 잘못된 생성자 테스트
- getScore 메소드 테스트

seunghye218 committed:
fix(NumbersValidator): add validation '0'
